### PR TITLE
check duplicate props

### PR DESCRIPTION
**Pull request by nevermore3**

### CODE REVIEW

File: src/validator/TraversalValidator.cpp [hunk 1]
```diff
@@ -213,5 +214,34 @@ Expression* TraversalValidator::buildNStepLoopCondition(uint32_t steps) const {
...
+    }
+    for (auto &prop : inputProps) {
+        DCHECK_NE(prop, "*");
+        NG_RETURN_IF_ERROR(checkPropNonexistOrDuplicate(inputs_, prop, validatorName));
+    }
+    return Status::OK();
+}
+
+Status TraversalValidator::checkVarProps(const std::string& validatorName) const {
+    auto& varProps = const_cast<ExpressionProps*>(&exprProps_)->varProps();
```
Comment (yixinglu) [inline src/validator/TraversalValidator.cpp:231]: why not add `const` qualifier for Expression::varProps/inputProps function?

File: src/validator/Validator.cpp [hunk 2]
```diff
@@ -372,5 +352,34 @@ StatusOr<std::string> Validator::checkRef(const Expression* ref, Value::Type typ
     }
 }
 
+Status Validator::checkDuplicateColName() {
```
Comment (Shylock-Hg) [inline src/validator/Validator.cpp:355]: I think you could check duplicate input in `Validator::checkRef`.

**Review by CPWstatic — DISMISSED** [review]

**Comment by yixinglu** [comment]
Maybe we need to rethink the userDefinedVarNameList_ usage.

**Review by CPWstatic — APPROVED** [review]

File: src/validator/Validator.cpp [hunk 3]
```diff
@@ -309,7 +309,7 @@ StatusOr<Value::Type> Validator::deduceExprType(const Expression* expr) const {
 }
 
 Status Validator::deduceProps(const Expression* expr, ExpressionProps& exprProps) {
-    DeducePropsVisitor visitor(qctx_, space_.id, &exprProps);
+    DeducePropsVisitor visitor(qctx_, space_.id, &exprProps, &userDefinedVarNameList_);
```
Comment (yixinglu) [inline src/validator/Validator.cpp:312]: This implies that we must check duplicate columns after props deduction. I think it's not proper to ensure the order in each subclass of validator.

File: src/validator/Validator.h [hunk 4]
```diff
@@ -106,7 +102,10 @@ class Validator {
 
     // Check the variable or input property reference
     // return the input variable
-    StatusOr<std::string> checkRef(const Expression *ref, const Value::Type type) const;
+    StatusOr<std::string> checkRef(const Expression *ref, const Value::Type type);
+
+    // Check the output for duplicate column names
+    Status checkDuplicateColName();
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply src/validator/Validator.h:135]: will change member variable in checkRef, so delete const qualifier

**Review by CPWstatic — DISMISSED** [review]

**Review by laura-ding — DISMISSED** [review]

**Review by jievince — APPROVED** [review]

File: src/validator/test/QueryValidatorTest.cpp [hunk 5]
```diff
@@ -1089,6 +1089,18 @@ TEST_F(QueryValidatorTest, GoInvalid) {
...
+                            "FROM $-.id OVER like";
+        auto result = checkResult(query);
+        EXPECT_EQ(std::string(result.message()), "SemanticError: Duplicate Column Name : `id'");
+    }
+    {
+        std::string query = "$a = GO FROM \"1\" OVER like YIELD like._dst AS id, like._src AS id; "
+                            "GO FROM $a.id OVER like";
+        auto result = checkResult(query);
+        EXPECT_EQ(std::string(result.message()), "SemanticError: Duplicate Column Name : `id'");
+    }
```
Comment (Shylock-Hg) [inline src/validator/test/QueryValidatorTest.cpp:1103]: Add a case about define duplicate alias but not used in latter sentence.

Comment (nevermore3) [reply src/validator/test/QueryValidatorTest.cpp:1103]: ok, will add case in next pr

File: src/validator/Validator.cpp [hunk 6]
```diff
@@ -375,5 +359,35 @@ Status Validator::toPlan() {
     VLOG(1) << "root: " << root_->kind() << " tail: " << tail_->kind();
     return Status::OK();
 }
+
+Status Validator::checkDuplicateColName() {
+    auto checkColName = [] (const ColsDef& nameList) {
+        std::unordered_map<std::string, bool> names;
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply src/validator/Validator.cpp:365]: ok

File: src/validator/test/QueryValidatorTest.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply src/validator/test/QueryValidatorTest.cpp:1103]: Ok.

File: src/validator/test/QueryValidatorTest.cpp [hunk 7]
```diff
@@ -1101,6 +1101,19 @@ TEST_F(QueryValidatorTest, GoInvalid) {
...
     }
+    {
+        std::string query = "GO FROM \"1\" OVER like, serve YIELD like._dst AS id, serve._src AS "
+                            "id, serve._dst AS DST | GO FROM $-.DST OVER like";
+        auto result = checkResult(query);
+        EXPECT_EQ(std::string(result.message()), "SemanticError: Duplicate Column Name : `id'");
+    }
+    {
+        std::string query =
+            "$a = GO FROM \"1\" OVER * YIELD like._dst AS id, like._src AS id, serve._dst as DST; "
```
Comment (Shylock-Hg) [inline src/validator/test/QueryValidatorTest.cpp:1112]: IMO, According to the discussion if the duplicate alias is not used won't report error.

Comment (nevermore3) [reply src/validator/test/QueryValidatorTest.cpp:1112]: emm, report error even if the duplicate alias is not used

Comment (Shylock-Hg) [reply src/validator/test/QueryValidatorTest.cpp:1112]: Ok, if so need to add case about that two variable one with duplicate alias the other without and the latter sentence use the variable without duplicate only.

**Review by laura-ding — APPROVED** [review]

**Review by jievince — APPROVED** [review]

File: src/validator/Validator.h [hunk 8]
```diff
@@ -147,6 +150,10 @@ class Validator {
     // root and tail of a subplan.
     PlanNode*                       root_{nullptr};
     PlanNode*                       tail_{nullptr};
+
+    ExpressionProps                 exprProps_;
+    // user define Variable name list
+    std::set<std::string>           userDefinedVarNameList_;
```
Comment (yixinglu) [inline src/validator/Validator.h:156]: unordered_set

**Pull request merged**